cpp/node_similarity_module/node_similarity_module.cpp

```diff
@@ -0,0 +1,151 @@
+    mgp::memory = memory;
+    const auto record_factory = mgp::RecordFactory(result);
+    const auto &arguments = mgp::List(args);
+    try {
+        insert_results(node_similarity_algs::CalculateSimilarityPairwise(arguments[0].ValueList(), arguments[1].ValueList(), node_similarity_util::Similarity::jaccard), record_factory);
+    } catch (const mgp::ValueException &e) {
+        record_factory.SetErrorMessage(e.what());
+    }
+}
+
+/*
+Calculates overlap similarity between given pairs of nodes.
+*/
+void OverlapPairwise(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+    mgp::memory = memory;
+    const auto &arguments = mgp::List(args);
+    const auto record_factory = mgp::RecordFactory(result);
+    try {
+        insert_results(node_similarity_algs::CalculateSimilarityPairwise(arguments[0].ValueList(), arguments[1].ValueList(), node_similarity_util::Similarity::overlap), record_factory);
+    } catch (const mgp::ValueException &e) {
+        record_factory.SetErrorMessage(e.what());
+    }
+}
+
+/*
+Calculates Jaccard similarity between all pairs of nodes.
+Jacc. similarity of two nodes can be calculated as len(intersection(neighbours(node1), neighbours(node2))) / len(union(neighbours(node1), neighbours(node2))) 
+*/
+void Jaccard(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+    mgp::memory = memory;
+    const auto record_factory = mgp::RecordFactory(result);
+    try {
+        insert_results(node_similarity_algs::CalculateSimilarityCartesian(mgp::Graph(memgraph_graph), node_similarity_util::Similarity::jaccard), record_factory);
+    } catch (const mgp::ValueException &e) {
+        record_factory.SetErrorMessage(e.what());
+    }
+}
+
+/*
+Calculates overlap similarity between all pairs of nodes.
+Overlap similarity of two nodes can be calculated as len(intersection(neighbours(node1), neighbours(node2))) / min(len(neighbours(node1), len(node2))) 
+*/
+void Overlap(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+    mgp::memory = memory;
+    const auto record_factory = mgp::RecordFactory(result);
+    try {
+        insert_results(node_similarity_algs::CalculateSimilarityCartesian(mgp::Graph(memgraph_graph), node_similarity_util::Similarity::overlap), record_factory);
+    } catch (const mgp::ValueException &e) {
+        record_factory.SetErrorMessage(e.what());
+    }
+
+}
+
+/*
+Calculates cosine similarity between all pairs of nodes.
+*/
+void Cosine(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+    mgp::memory = memory;
+    const auto record_factory = mgp::RecordFactory(result);
+    const auto &arguments = mgp::List(args);
+    try {
+        insert_results(node_similarity_algs::CalculateSimilarityCartesian(mgp::Graph(memgraph_graph), node_similarity_util::Similarity::cosine, arguments[0].ValueString().data()), record_factory);
```

Comment:
Why does this function insert different result fields than the analogous `Jaccard` and `Overlap` functions?

Reply:
All fields are the same: `arguments[0].ValueString()` is sent to calculate similarity cartesian func

Reply:
I see - however, why do we need to pass `arguments[0].ValueString()`? `Jaccard` and `Overlap` are analogous to `Cosine` yet they don’t take a string argument. Is it possible to remove it so that the signature of `Cosine` is consistent with them?

Reply:
Please take a look at the implementation. Cosine func works on properties so we need property name

Reply:
Resolved